client/context/broadcast.go

```diff
@@ -103,23 +103,21 @@ func (ctx CLIContext) broadcastTxAsync(txBytes []byte) (*ctypes.ResultBroadcastT
 		return res, err
 	}
 
-	if ctx.Output != nil {
-		if ctx.OutputFormat == "json" {
-			type toJSON struct {
-				TxHash string
-			}
+	if ctx.OutputIsSet() && ctx.OutputIsJSON() {
```

Comment:
* I see how `ctx.OutputIsSet()` is more helpful, but honestly I think it might be overkill. Checking if it's nil should suffice.
* What I had in mind was something along the following:

```go
type TxBroadcastResponse struct {
    TxHash string                  `json:"tx_hash,omitempty"`
    Height uint64                  `json:"height,omitempty"`
    Result ResultBroadcastTxString `json:"result,omitempty"`
}

func (txbr TxBroadcastResponse) String() string {
    // ...
}
```

All the broadcast methods, `BroadcastTx`, `BroadcastTxSync`, and `BroadcastTxAsync`, will return `(TxBroadcastResponse, error)`. They will populate any fields as necessary and anything else can be omitted.

The REST handler, `BroadcastTxRequest`, can simply return this `TxBroadcastResponse` JSON.

The CLI handler, `CompleteAndBroadcastTxCLI`, will take the `TxBroadcastResponse` result and call `cliCtx.PrintOutput`.

Does this make sense? This way we don't need to have any `ctx.OutputIsJSON()` or `ctx.Output.Write(bz)` calls. All of this can be much cleaner.

I haven't fully gone through the flow, but looking at the code, I don't see why this can't work.

Reply:
I've refactored to match this type.